ddsrecorder/src/cpp/command_receiver/CommandReceiver.cpp

```diff
@@ -0,0 +1,298 @@
+        }
+        if (status_topic_ != nullptr)
+        {
+            participant_->delete_topic(status_topic_);
+        }
+        if (command_subscriber_ != nullptr)
+        {
+            if (command_reader_ != nullptr)
+            {
+                command_subscriber_->delete_datareader(command_reader_);
+            }
+            participant_->delete_subscriber(command_subscriber_);
+        }
+        if (status_publisher_ != nullptr)
+        {
+            if (status_writer_ != nullptr)
+            {
+                status_publisher_->delete_datawriter(status_writer_);
+            }
+            participant_->delete_publisher(status_publisher_);
+        }
+        DomainParticipantFactory::get_instance()->delete_participant(participant_);
+    }
+}
+
+void CommandReceiver::wait_for_command()
+{
+    command_received_.store(CommandCode::NONE);
+    event_handler_->wait_for_event();
+    event_handler_->reset_event_count();
+}
+
+CommandCode CommandReceiver::command_received()
+{
+    return command_received_.load();
+}
+
+void CommandReceiver::publish_status(
+        CommandCode current,
+        CommandCode previous,
+        std::string info)
+{
+    Status status;
+    status.current(command_to_status_string_(current));
+    status.previous(command_to_status_string_(previous));
+    if (!info.empty())
+    {
+        status.info(info);
+    }
+    logInfo(
+        DDSRECORDER_COMMAND_RECEIVER,
+        "Publishing status: " << status.previous() << " ---> " << status.current() <<  " with info [" << status.info() <<
+            " ].");
+    status_writer_->write(&status);
+}
+
+void CommandReceiver::on_subscription_matched(
+        DataReader*,
+        const SubscriptionMatchedStatus& info)
+{
+    if (info.current_count_change == 1)
+    {
+        logInfo(
+            DDSRECORDER_COMMAND_RECEIVER,
+            "Subscriber matched [ " << iHandle2GUID(info.last_publication_handle) << " ].");
+    }
+    else if (info.current_count_change == -1)
+    {
+        logInfo(
+            DDSRECORDER_COMMAND_RECEIVER,
+            "Subscriber unmatched [ " << iHandle2GUID(info.last_publication_handle) << " ].");
+    }
+    else
+    {
+        logWarning(
+            DDSRECORDER_COMMAND_RECEIVER,
+            info.current_count_change << " is not a valid value for SubscriptionMatchedStatus current count change");
+    }
+}
+
+void CommandReceiver::on_data_available(
+        DataReader* reader)
+{
+    SampleInfo info;
+    ControllerCommand controller_command;
+    while ((reader->take_next_sample(&controller_command,
+            &info)) == (ReturnCode_t::RETCODE_OK && info.instance_state == ALIVE_INSTANCE_STATE))
+    {
+        std::string command = controller_command.command();
+        std::string args = controller_command.args();
+        logInfo(
+            DDSRECORDER_COMMAND_RECEIVER,
+            "New command received: " << command << " [" << args << "]");
+
+        if (command_received_ == CommandCode::CLOSE)
```

Comment:
Is this supposed to be here? command_received_ is not set yet by the last message

Reply:
I think so, once a CLOSE command is received there is no way back, it would make no sense to keep reacting I believe.

Reply:
The thing is that is compared to close before set

Reply:
and shouldn't it have also a `simulate_event_occurred`?

Reply:
When the receiver receives a new command, the first thing that does is to verify the previous command was not CLOSE. If it was, the recorder (tool process) is in the process of terminating, and so there is no need to keep processing any more commands.